File: 17.js
// part one
(() => {
  const registers = {
    A: 32916674,
    B: 0,
    C: 0,
  };

  const codes = [2,4,1,1,7,5,0,3,1,4,4,0,5,5,3,0];

  const combos = {
    0: () => 0,
    1: () => 1,
    2: () => 2,
    3: () => 3,
    4: () => registers.A,
    5: () => registers.B,
    6: () => registers.C,
  };

  const output = [];

  for (let i = 0; i < codes.length; ) {
    const operations = {
      0: (val) => {
        registers.A = Math.floor(registers.A / Math.pow(2, combos[val]()));
        i += 2;
      },
      1: (val) => {
        registers.B = registers.B ^ val;
        i += 2
      },
      2: (val) => {
        registers.B = combos[val]() % 8;
        i += 2
      },
      3: (val) => {
        if (registers.A === 0) {
          i += 2;
          return;
        }
        if (i !== val) {
          i = val;
        } else {
          i += 2;
        }
      },
      4: () => {
        registers.B = registers.B ^ registers.C;
        i += 2;
      },
      5: (val) => {
        output.push(combos[val]() % 8);
        i += 2;
      },
      6: (val) => {
        registers.B = Math.floor(registers.A / Math.pow(2, combos[val]()));
        i += 2;
      },
      7: (val) => {
        registers.C = Math.floor(registers.A / Math.pow(2, combos[val]()));
        i += 2;
      }
    }

    operations[codes[i]](codes[i + 1]);
  }

  return output.join(',');
})();

// part two done in python

File: 17.py
"""
part two

Found some annoying behavior with large numbers and bitwise XOR with JS, leading to negative results.
Lost confidence in JS and switched to Python.

- Found the following are true about my input:
  - There's only one go to, at the very end
  - B and C are purely ephemeral state and don't carry over to the next loop
  - A is continually divided by 8
  - A can affect the next result -- but it doesn't always
- Manually did some searching to find how to construct blocks of the input
- Some of the blocks would affect "lower" bits, so I made sure to:
  - Only consider the "lowest" result for a block
  - Do not use that block if it's the kind that will affect "lower" bits
- Thankfully I found some groupings fro which that worked
- There were still 6 codes at the beginning to figure out, so I just manually iterated
  over the remaining possibilities from there.
"""

def get_output(A):
  return ((A % 8) ^ 1 ^ 4 ^ (A // 2**((A % 8) ^ 1))) % 8

def solve(A):
  out = []
  curr = A
  while curr != 0:
    out.append(get_output(curr))
    curr //= 8
  return out

goal = [2,4,1,1,7,5,0,3,1,4,4,0,5,5,3,0]

want = [
  goal[:4],
  goal[4:8],
  goal[8:12],
  goal[12:],
  [0,3,1,4,4,0],
  [4,1,1,7,5]
]

# next line was found from searching for the "want" blocks above
# and then multiplying them "up" to become the "higher bits" so
# we're left with just the "lower" bits missing
pretty_good = 2944 * (8 ** 12) + 177646 * (8 ** 6)

for i in range(pretty_good, pretty_good + 8 ** 6):
  if i % 10000 == 0:
    print(i - pretty_good, 'of', pretty_good + 8 ** 6 - i)
  res = solve(i)
  if res == goal:
    print(i, res)
